File: TiAires/app/src/main/java/net/teamtruta/tiaires/GeoCacheListAdapter.java
package net.teamtruta.tiaires;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.text.Spanned;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.GridLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;
import androidx.core.text.HtmlCompat;
import androidx.recyclerview.widget.RecyclerView;

import net.teamtruta.tiaires.data.GeoCache;
import net.teamtruta.tiaires.data.GeoCacheAttribute;
import net.teamtruta.tiaires.data.GeoCacheInTour;
import net.teamtruta.tiaires.data.GeoCacheInTourWithDetails;
import net.teamtruta.tiaires.data.GeoCacheLog;
import net.teamtruta.tiaires.data.GeoCacheWithLogsAndAttributes;
import net.teamtruta.tiaires.viewModels.TourViewModel;

import org.jetbrains.annotations.NotNull;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;


class GeoCacheListAdapter extends RecyclerView.Adapter<GeoCacheListAdapter.ViewHolder> implements ItemTouchHelperAdapter{

    private static List<GeoCacheInTourWithDetails> geoCacheInTourList = new ArrayList();
    private final EditOnClickListener editOnClickListener;
    private final GoToOnClickListener goToOnClickListener;
    private final Context context;
    static TourViewModel viewModel;

    static OnVisitListener onVisitListener;


    GeoCacheListAdapter(EditOnClickListener editOnClickListener,
                        GoToOnClickListener goToOnClickListener,
                        Context applicationContext, TourViewModel viewModel){
        this.editOnClickListener = editOnClickListener;
        this.goToOnClickListener = goToOnClickListener;
        this.context = applicationContext;
        this.viewModel = viewModel;
    }

    public void setGeoCacheInTourList(List<GeoCacheInTourWithDetails> list){
        this.geoCacheInTourList = list;
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount(){
        return geoCacheInTourList.size();
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, final int position){

        GeoCacheInTour geoCacheInTour = geoCacheInTourList.get(position).getGeoCacheInTour();
        GeoCacheWithLogsAndAttributes geoCacheWithLogsAndAttributes = geoCacheInTourList.get(position).getGeoCache();
        GeoCache geoCache = geoCacheWithLogsAndAttributes.getGeoCache();

        // 1. Set geoCache name
        holder.geoCacheName.setText(geoCache.getName());

        // 2. Set geoCache type symbol
        int drawableID;

        if(geoCacheInTour.getCurrentVisitOutcome() == VisitOutcomeEnum.Found){
            drawableID = R.drawable.geo_cache_icon_found;
        } else if(geoCacheInTour.getCurrentVisitOutcome() == VisitOutcomeEnum.DNF){
            drawableID = R.drawable.geo_cache_icon_dnf;
        } else if(geoCacheInTour.getCurrentVisitOutcome() == VisitOutcomeEnum.Disabled){
            drawableID = R.drawable.geo_cache_icon_disabled;
        } else {
            drawableID = GeoCacheIcon.Companion.getIconDrawable(geoCache.getType());
        }

        Drawable geoCacheSymbolDrawable = ContextCompat.getDrawable(holder.view.getContext(), drawableID);
        holder.geoCacheSymbol.setImageDrawable(geoCacheSymbolDrawable);

        // 3. Set information line 1: Code, difficulty and terrain and Size
        holder.geoCacheCode.setText(geoCache.getCode());

        String difTerString = context.getString(R.string.geo_cache_dif_ter);
        double difficulty = geoCache.getDifficulty();
        double terrain = geoCache.getTerrain();
        
        String diffString = String.format(Locale.getDefault(), "%.1f", difficulty);
        String terString = String.format(Locale.getDefault(), "%.1f", terrain);
        diffString = ((difficulty > 4) ?
                "<font color='red'>" + diffString + "</font>/" : diffString);
        terString = ((terrain > 4) ?
                "<font color='red'>" + terString + "</font>/" : terString);
        holder.geoCacheDifTer.setText( HtmlCompat.fromHtml(String.format(difTerString, diffString, terString),
                HtmlCompat.FROM_HTML_MODE_LEGACY));

        String sizeString = context.getString(R.string.geo_cache_size);
        holder.geoCacheSize.setText(String.format(sizeString, geoCache.getSize()));


        // 4. Set information line 2: Favorites and whether geoCache has hint
        String favString = context.getString(R.string.geo_cache_favs);
        holder.geoCacheFavs.setText(String.format(favString, geoCache.getFavourites()));
        if(geoCache.hasHint()){
            String hintString = context.getString(R.string.geo_cache_has_hint);
            holder.geoCacheHasHint.setText(HtmlCompat.fromHtml(
                    hintString,
                    HtmlCompat.FROM_HTML_MODE_LEGACY));
        } else {
            holder.geoCacheHasHint.setText(HtmlCompat.fromHtml(context.getString(R.string.geo_cache_has_no_hint),
                    HtmlCompat.FROM_HTML_MODE_LEGACY));
        }

        List<GeoCacheLog> last10Logs = geoCacheWithLogsAndAttributes.getLastNLogs(10);
        int i = 0;
        for (GeoCacheLog log:last10Logs) {
            TextView tv = (TextView) holder.lastLogsLayout.getChildAt(i);
            if(log.getLogType() == VisitOutcomeEnum.Found){
                tv.setTextColor(holder.view.getContext().getColor(R.color.colorPrimary));
            } else if( log.getLogType() == VisitOutcomeEnum.DNF ){
                tv.setTextColor(holder.view.getContext().getColor(R.color.red));
            } else {
                tv.setTextColor(holder.view.getContext().getColor(R.color.blue));
            }
            i++;
        }


        // 5. GeoCache Hint
        holder.hint.setText(getHintText(geoCache));


        // 6. Set DNF information if required
        if(geoCacheWithLogsAndAttributes.isDNFRisk()){
            String dnfString = context.getString(R.string.dnf_risk);
            dnfString = String.format(dnfString, geoCacheWithLogsAndAttributes.getDNFRisk());

            holder.dnfInfo.setText(HtmlCompat.fromHtml(
                    dnfString,
                    HtmlCompat.FROM_HTML_MODE_LEGACY));
            holder.dnfInfo.setVisibility(View.VISIBLE);
        } else {
            holder.dnfInfo.setText("");
            holder.dnfInfo.setVisibility(View.GONE);
        }

       // 7. Set listener for edit button
        holder.editButton.setOnClickListener(v -> {
            if(editOnClickListener!=null){
                editOnClickListener.onEditClick(geoCacheInTour.getId());
            }
        });

        // 8. Set listener for go-to-geoCache button
        holder.goToButton.setOnClickListener(v -> {
            if(goToOnClickListener!=null){
                goToOnClickListener.onGoToClick(geoCache);
            }
        });

        // 9. Make section grey if geoCache has been visited
        if(geoCacheInTour.getCurrentVisitOutcome() == VisitOutcomeEnum.Found
                || geoCacheInTour.getCurrentVisitOutcome() == VisitOutcomeEnum.DNF
                || geoCacheInTour.getCurrentVisitOutcome() == VisitOutcomeEnum.Disabled){
            holder.layout.setBackgroundColor(context.getColor(R.color.light_grey));
        } else {
            holder.layout.setBackgroundColor(context.getColor(R.color.white));
        }

        // 10. Setup expansion
        holder.extraInfoArrow.setOnClickListener(v -> holderExpansionOnClicklistener(holder, geoCacheWithLogsAndAttributes));

        // 11. Add Cache Attributes
        holder.attributeList.removeAllViews();
        if(geoCacheWithLogsAndAttributes.getAttributes().size() > 0){
            for(GeoCacheAttribute attribute: geoCacheWithLogsAndAttributes.getAttributes()){
                ImageView iv = new ImageView(context);
                int attributeIconID = GeoCacheAttributeIcon.Companion.getGeoCacheAttributeIcon(attribute.getAttributeType());
                iv.setImageDrawable(ContextCompat.getDrawable(context, attributeIconID));
                // DEPRECATED:
                //iv.setImageDrawable(App.getContext().getDrawable(attributeIconID));
                LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(
                        (int) context.getResources().getDimension(R.dimen.medium_icon_size),
                        (int) context.getResources().getDimension(R.dimen.medium_icon_size));
                layoutParams.setMarginEnd((int) context.getResources().getDimension(R.dimen.tiny_padding));
                iv.setLayoutParams(layoutParams);
                holder.attributeList.addView(iv);
            }
        }
    }


    private Spanned getHintText(GeoCache geoCache) {

        String hintString = "<strong>HINT</strong>: <i>" + geoCache.getHint() + "</i>";
        return HtmlCompat.fromHtml(hintString, HtmlCompat.FROM_HTML_MODE_LEGACY);

    }

    void holderExpansionOnClicklistener(ViewHolder holder, GeoCacheWithLogsAndAttributes geoCacheWithLogsAndAttributes){
        if(holder.extraInfoArrow.isChecked() || holder.extraInfoLayout.getVisibility() != View.VISIBLE){
            expandHolder(holder, geoCacheWithLogsAndAttributes);
        } else {
            unexpandHolder(holder);
        }
    }
    void expandHolder(ViewHolder holder, GeoCacheWithLogsAndAttributes geoCacheWithLogsAndAttributes){

        // Remove hint indication from line 1
        if(geoCacheWithLogsAndAttributes.getGeoCache().hasHint()){
            holder.geoCacheHasHint.setVisibility(View.GONE);
            holder.hint.setVisibility(View.VISIBLE);
        }

        // Show extra information
        holder.extraInfoLayout.setVisibility(View.VISIBLE);

        holder.extraInfoArrow.setChecked(true);

        // Show attributes
        if(geoCacheWithLogsAndAttributes.getAttributes().size() != 0)
            holder.attributeList.setVisibility(View.VISIBLE);

    }

    void unexpandHolder(ViewHolder holder){
        // Hide extra information
        holder.geoCacheHasHint.setVisibility(View.VISIBLE);
        holder.hint.setVisibility(View.GONE);

        holder.extraInfoLayout.setVisibility(View.GONE);
        holder.extraInfoArrow.setChecked(false);

        // Hide attributes
        holder.attributeList.setVisibility(View.GONE);
    }

    @NotNull
    public GeoCacheListAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType){

        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.element_geo_cache_layout, parent, false);
        return new ViewHolder(v);
    }

    static void visitItem(int position, VisitOutcomeEnum visit){

        // Save the deleted item in case user wants to undo the action;

        // Useful if we want to add an undo functionality here
        // recentlyVisitedCache = tour.getCacheInTour(position);
        // recentlyVisitedCachePosition = position;

        GeoCacheInTour selectedGeoCache = geoCacheInTourList.get(position).getGeoCacheInTour();
        selectedGeoCache.setCurrentVisitOutcome(visit);
        selectedGeoCache.setCurrentVisitDatetime(Instant.now());
        viewModel.updateGeoCacheInTour(selectedGeoCache);

        onVisitListener.onVisit(visit.toString());
    }

    //@Override
    //public void onItemDismiss(int position) {
        //tour._tourCaches.remove(position);
        //notifyItemRemoved(position);
    //}

    @Override
    public void onItemMove(int fromPosition, int toPosition) {

        Collections.swap(geoCacheInTourList, fromPosition, toPosition);
        notifyItemMoved(fromPosition, toPosition);
        /*
        tour.swapCachePositions(fromPosition, toPosition);
        tour.updateTourCaches();*/

    }

    public void onMoveEnded(){
        /*tour.updateTourCaches();*/
    }

    static class ViewHolder extends RecyclerView.ViewHolder{

        View view;
        TextView geoCacheName;
        ImageView geoCacheSymbol;
        TextView geoCacheCode;
        TextView geoCacheDifTer;
        TextView geoCacheSize;
        TextView geoCacheFavs;
        TextView geoCacheHasHint;
        TextView dnfInfo;
        TextView dnfInfoExpanded;
        TextView hint;
        Button editButton;
        Button goToButton;
        ConstraintLayout layout;
        ConstraintLayout extraInfoLayout;
        CheckBox extraInfoArrow;
        GridLayout lastLogsLayout;
        LinearLayout attributeList;

        ViewHolder(View v){
            super(v);
            view = v;

            geoCacheName = v.findViewById(R.id.geo_cache_title);
            geoCacheSymbol = v.findViewById(R.id.geo_cache_symbol);
            geoCacheCode = v.findViewById(R.id.geo_cache_code);
            geoCacheDifTer = v.findViewById(R.id.geo_cache_dif_ter);
            geoCacheSize = v.findViewById(R.id.geo_cache_size);
            geoCacheFavs = v.findViewById(R.id.geo_cache_favs);
            geoCacheHasHint = v.findViewById(R.id.geo_cache_has_hint);
            dnfInfo = v.findViewById(R.id.dnf_risk);
            dnfInfoExpanded = v.findViewById(R.id.dnf_info_expanded);
            hint = v.findViewById(R.id.hint);
            editButton = v.findViewById(R.id.edit_button);
            goToButton = v.findViewById(R.id.go_to_button);
            layout = v.findViewById(R.id.element_geo_cache_layout);
            extraInfoLayout = v.findViewById(R.id.expandable_info);
            extraInfoArrow = v.findViewById(R.id.extra_info_arrow);
            lastLogsLayout = v.findViewById(R.id.last10LogsSquares);
            attributeList = v.findViewById(R.id.geo_cache_attributes);
        }

    }

    interface EditOnClickListener {
        void onEditClick(long geoCacheID);
    }

    interface GoToOnClickListener{
        void onGoToClick(GeoCache geoCache);
    }

    interface OnVisitListener{
        void onVisit(String visit);
    }
}

File: TiAires/app/src/main/java/net/teamtruta/tiaires/adapters/GeoCacheListAdapter.kt
package net.teamtruta.tiaires.adapters

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Typeface
import android.text.Spanned
import android.text.style.ForegroundColorSpan
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import net.teamtruta.tiaires.App
import net.teamtruta.tiaires.R
import net.teamtruta.tiaires.data.models.*
import net.teamtruta.tiaires.extensions.GeoCacheAttributeIcon.Companion.getGeoCacheAttributeIcon
import net.teamtruta.tiaires.extensions.GeoCacheIcon.Companion.getIconDrawable
import net.teamtruta.tiaires.viewModels.TourViewModel
import net.teamtruta.tiaires.viewModels.TourViewModelFactory
import net.teamtruta.tiaires.views.TourActivity
import java.time.Instant
import java.util.*
import android.text.SpannableString
import android.text.SpannableStringBuilder
import android.text.style.StyleSpan


class GeoCacheListAdapter(private val editOnClickListener: EditOnClickListener?,
                          private val goToOnClickListener: GoToOnClickListener?,
                          private val viewHolderOnClickListener: ViewHolder.ClickListener,
                          private val context: Context,
                          private val viewModel: TourViewModel,
                          val activity: TourActivity) :
        RecyclerView.Adapter<GeoCacheListAdapter.ViewHolder>(), ItemTouchHelperAdapter {

    fun setGeoCacheInTourList(list: List<GeoCacheInTourWithDetails?>) {
        geoCacheInTourList = list
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return geoCacheInTourList.size
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val geoCacheInTourWithDetails = geoCacheInTourList[position]
        val geoCacheInTour = geoCacheInTourWithDetails!!.geoCacheInTour
        val geoCacheWithLogsAndAttributes = geoCacheInTourList[position]!!.geoCache
        val geoCache = geoCacheWithLogsAndAttributes.geoCache

        // 1. Set geoCache name
        holder.geoCacheName.text = geoCache.name

        // 2. Set geoCache type symbol
        val drawableID: Int = if (geoCacheInTour.currentVisitOutcome === VisitOutcomeEnum.Found) {
            R.drawable.geo_cache_icon_found
        } else if (geoCacheInTour.currentVisitOutcome === VisitOutcomeEnum.DNF) {
            R.drawable.geo_cache_icon_dnf
        } else if (geoCacheInTour.currentVisitOutcome === VisitOutcomeEnum.Disabled) {
            R.drawable.geo_cache_icon_disabled
        } else {
            getIconDrawable(geoCache.type)
        }
        val geoCacheSymbolDrawable = ContextCompat.getDrawable(holder.view.context, drawableID)
        holder.geoCacheSymbol.setImageDrawable(geoCacheSymbolDrawable)

        // 3. Set information line 1: Code, difficulty and terrain and size
        val red = ForegroundColorSpan(context.getColor(R.color.red))
        holder.geoCacheCode.text = geoCache.code
        val difficulty = geoCache.difficulty
        val terrain = geoCache.terrain

        val difficultyTerrainString = SpannableString(String.format(context.getString(R.string.geo_cache_dif_ter), difficulty, terrain))
        if (difficulty > 4) difficultyTerrainString.setSpan(red, 5, 8, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        if (terrain > 24) difficultyTerrainString.setSpan(red, 9, 12, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        holder.geoCacheDifTer.text = difficultyTerrainString

        val sizeString = context.getString(R.string.geo_cache_size)
        holder.geoCacheSize.text = String.format(sizeString, geoCache.size)


        // 4. Set information line 2: Favorites and whether geoCache has hint
        val favString = context.getString(R.string.geo_cache_favs)
        holder.geoCacheFavs.text = String.format(favString, geoCache.favourites)
        if (geoCache.hasHint()) {
            val hintString = " - " + context.getString(R.string.geo_cache_has_hint)
            holder.geoCacheHasHint.text = hintString
        } else {
            val hintString = " - " + context.getString(R.string.geo_cache_has_no_hint)
            val string = SpannableString(hintString)
            string.setSpan(red, 3, string.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            holder.geoCacheHasHint.text = string
        }
        val last10Logs = geoCacheWithLogsAndAttributes.getLastNLogs(10)
        var i = 0
        for ((_, logType) in last10Logs) {
            val tv = holder.lastLogsLayout.getChildAt(i) as TextView
            when {
                logType === VisitOutcomeEnum.Found -> {
                    tv.setTextColor(holder.view.context.getColor(R.color.colorPrimary))
                }
                logType === VisitOutcomeEnum.DNF -> {
                    tv.setTextColor(holder.view.context.getColor(R.color.red))
                }
                else -> {
                    tv.setTextColor(holder.view.context.getColor(R.color.blue))
                }
            }
            i++
        }


        // 5. GeoCache Hint
        holder.hint.text = getHintText(geoCache)


        // 6. Set DNF information if required
        if (geoCacheWithLogsAndAttributes.isDNFRisk()) {

            val dnfString = SpannableString(" - " + geoCacheWithLogsAndAttributes.dNFRisk)
            dnfString.setSpan(red, 3, dnfString.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            holder.dnfInfo.text = dnfString
            holder.dnfInfo.visibility = View.VISIBLE
        } else {
            holder.dnfInfo.text = ""
            holder.dnfInfo.visibility = View.GONE
        }

        // 7. Set listener for edit button
        holder.editButton.setOnClickListener {
            editOnClickListener?.onEditClick(geoCacheInTourList[position]!!)
        }

        // 8. Set listener for go-to-geoCache button
        holder.goToButton.setOnClickListener {
            goToOnClickListener?.onGoToClick(geoCache)
        }

        // 9. Make section grey if geoCache has been visited
        if (geoCacheInTour.currentVisitOutcome === VisitOutcomeEnum.Found || geoCacheInTour.currentVisitOutcome === VisitOutcomeEnum.DNF || geoCacheInTour.currentVisitOutcome === VisitOutcomeEnum.Disabled) {
            holder.layout.setBackgroundColor(context.getColor(R.color.light_grey))
        } else {
            holder.layout.setBackgroundColor(context.getColor(R.color.white))
        }

        // 10. Setup expansion
        holder.extraInfoArrow.setOnClickListener {holderExpansionOnClicklistener(holder, geoCacheWithLogsAndAttributes) }

        // 11. Add Cache Attributes
        holder.attributeList.removeAllViews()
        if (geoCacheWithLogsAndAttributes.attributes.isNotEmpty()) {
            for ((_, attributeType) in geoCacheWithLogsAndAttributes.attributes) {
                val iv = ImageView(context)
                val attributeIconID = getGeoCacheAttributeIcon(attributeType)
                iv.setImageDrawable(ContextCompat.getDrawable(context, attributeIconID))
                // DEPRECATED:
                //iv.setImageDrawable(App.getContext().getDrawable(attributeIconID));
                val layoutParams = LinearLayout.LayoutParams(
                        context.resources.getDimension(R.dimen.medium_icon_size).toInt(),
                        context.resources.getDimension(R.dimen.medium_icon_size).toInt())
                layoutParams.marginEnd = context.resources.getDimension(R.dimen.tiny_padding).toInt()
                iv.layoutParams = layoutParams
                holder.attributeList.addView(iv)
            }
        }

        // Setup Edit Mode
        holder.geoCacheSelectedCheckbox.setOnClickListener{
            viewHolderOnClickListener.onItemClicked(position)}

        if(TourActivity.isInActionMode){
            unexpandHolder(holder)
            holder.geoCacheSymbol.visibility = View.INVISIBLE
            holder.extraInfoArrow.visibility = View.INVISIBLE
            holder.editButton.visibility = View.GONE
            holder.goToButton.visibility = View.GONE
            holder.geoCacheSelectedCheckbox.visibility = View.VISIBLE
            holder.reorderHandle.visibility = View.VISIBLE
            val geocacheIsSelected = TourActivity.selectedGeoCacheCodes.contains(geoCache.code)
            holder.geoCacheSelectedCheckbox.isChecked = geocacheIsSelected
        } else {
            holder.geoCacheSymbol.visibility = View.VISIBLE
            holder.extraInfoArrow.visibility = View.VISIBLE
            holder.editButton.visibility = View.VISIBLE
            holder.goToButton.visibility = View.VISIBLE
            holder.geoCacheSelectedCheckbox.visibility = View.GONE
            holder.reorderHandle.visibility = View.GONE
        }

    }

    private fun getHintText(geoCache: GeoCache): Spanned {
        val stringBuilder = SpannableStringBuilder()
        stringBuilder.append("HINT: ", StyleSpan(Typeface.BOLD), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        stringBuilder.append(geoCache.hint, StyleSpan(Typeface.ITALIC), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        return stringBuilder
    }

    private fun holderExpansionOnClicklistener(holder: ViewHolder, geoCacheWithLogsAndAttributesAndWaypoints: GeoCacheWithLogsAndAttributesAndWaypoints) {
        if (holder.extraInfoArrow.isChecked || holder.extraInfoLayout.visibility != View.VISIBLE) {
            expandHolder(holder, geoCacheWithLogsAndAttributesAndWaypoints)
        } else {
            unexpandHolder(holder)
        }
    }

    private fun expandHolder(holder: ViewHolder, geoCacheWithLogsAndAttributesAndWaypoints: GeoCacheWithLogsAndAttributesAndWaypoints) {

        // Remove hint indication from line 1
        if (geoCacheWithLogsAndAttributesAndWaypoints.geoCache.hasHint()) {
            holder.geoCacheHasHint.visibility = View.GONE
            holder.hint.visibility = View.VISIBLE
        }

        // Show extra information
        holder.extraInfoLayout.visibility = View.VISIBLE
        holder.extraInfoArrow.isChecked = true

        // Show attributes
        if (geoCacheWithLogsAndAttributesAndWaypoints.attributes.isNotEmpty())
            holder.attributeList.visibility = View.VISIBLE
    }

    private fun unexpandHolder(holder: ViewHolder) {
        // Hide extra information
        holder.geoCacheHasHint.visibility = View.VISIBLE
        holder.hint.visibility = View.GONE
        holder.extraInfoLayout.visibility = View.GONE
        holder.extraInfoArrow.isChecked = false

        // Hide attributes
        holder.attributeList.visibility = View.GONE
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.element_geo_cache_layout, parent, false)
        val viewHolder =  ViewHolder(v, viewHolderOnClickListener)
        viewHolder.reorderHandle.setOnTouchListener{
            _, event -> if(event.actionMasked == MotionEvent.ACTION_DOWN){
                activity.startDragging(viewHolder)
                }
            return@setOnTouchListener true
        }
        return viewHolder
    }

    override fun onItemMove(fromPosition: Int, toPosition: Int) {
        Collections.swap(geoCacheInTourList, fromPosition, toPosition)
        notifyItemMoved(fromPosition, toPosition)
    }

    fun onMoveEnded() {
        viewModel.reorderTourCaches(geoCacheInTourList)
    }

    class ViewHolder(var view: View, private val onClickListener: ClickListener) :
        RecyclerView.ViewHolder(view), View.OnClickListener, View.OnLongClickListener {

        var geoCacheName: TextView = view.findViewById(R.id.geo_cache_title)
        var geoCacheSymbol: ImageView = view.findViewById(R.id.geo_cache_symbol)
        var geoCacheCode: TextView = view.findViewById(R.id.geo_cache_code)
        var geoCacheDifTer: TextView = view.findViewById(R.id.geo_cache_dif_ter)
        var geoCacheSize: TextView = view.findViewById(R.id.geo_cache_size)
        var geoCacheFavs: TextView = view.findViewById(R.id.geo_cache_favs)
        var geoCacheHasHint: TextView = view.findViewById(R.id.geo_cache_has_hint)
        var dnfInfo: TextView = view.findViewById(R.id.dnf_risk)
        var hint: TextView = view.findViewById(R.id.hint)
        var editButton: Button = view.findViewById(R.id.edit_button)
        var goToButton: Button = view.findViewById(R.id.go_to_button)
        var layout: ConstraintLayout = view.findViewById(R.id.element_geo_cache_layout)
        var extraInfoLayout: ConstraintLayout = view.findViewById(R.id.expandable_info)
        var extraInfoArrow: CheckBox = view.findViewById(R.id.extra_info_arrow)
        var lastLogsLayout: GridLayout = view.findViewById(R.id.last10LogsSquares)
        var attributeList: LinearLayout = view.findViewById(R.id.geo_cache_attributes)

        var geoCacheSelectedCheckbox: CheckBox = view.findViewById(R.id.geo_cache_selection_checkbox)
        var reorderHandle: ImageView = view.findViewById(R.id.reorder_handle)

        // Setup click listener so that long click enables action mode
        init {
            view.setOnLongClickListener(this)
            view.setOnClickListener(this)
            }


        override fun onLongClick(view: View?): Boolean {
            onClickListener.onItemLongClicked(absoluteAdapterPosition)
            return true
        }

        override fun onClick(view: View?) {
            onClickListener.onItemClicked(absoluteAdapterPosition)
        }

        interface ClickListener {
            fun onItemClicked(position: Int)
            fun onItemLongClicked(position: Int): Boolean
        }

    }

    interface EditOnClickListener {
        fun onEditClick(geoCacheInTour: GeoCacheInTourWithDetails)
    }

    interface GoToOnClickListener {
        fun onGoToClick(geoCache: GeoCache?)
    }

    interface OnVisitListener {
        fun onVisit(visit: String?)
    }



    companion object {
        private var geoCacheInTourList: List<GeoCacheInTourWithDetails?> = listOf()


        //var onVisitListener: OnVisitListener? = null

        fun visitItem(position: Int, visit: VisitOutcomeEnum) {

            // Save the deleted item in case user wants to undo the action;

            // Useful if we want to add an undo functionality here
            // recentlyVisitedCache = tour.getCacheInTour(position);
            // recentlyVisitedCachePosition = position;
            val selectedGeoCache = geoCacheInTourList[position]!!.geoCacheInTour
            selectedGeoCache.currentVisitOutcome = visit
            selectedGeoCache.currentVisitDatetime = Instant.now()

            val viewModel: TourViewModel = TourViewModelFactory(App().repository)
                    .create(TourViewModel::class.java)

            viewModel.updateGeoCacheInTour(selectedGeoCache)
        }
    }

}
